src/java/picard/vcf/MendelianViolationDetector.java

```diff
@@ -0,0 +1,338 @@
+                if (ctx.hasAttribute(VCFConstants.ALLELE_FREQUENCY_KEY))
+                    builder.attribute("AF_Orig", ctx.getAttribute(VCFConstants.ALLELE_FREQUENCY_KEY));
+                if (ctx.hasAttribute(VCFConstants.ALLELE_NUMBER_KEY))
+                    builder.attribute("AN_Orig", ctx.getAttribute(VCFConstants.ALLELE_NUMBER_KEY));
+
+                // Write out the variant record
+                familyToViolations.get(trio.FAMILY_ID).add(builder.make());
+            }
+        }
+    }
+
+    @Override
+    public Result result() {
+        return new Result(trios, familyToViolations);
+    }
+
+    /** Tests to ensure that a locus is bi-allelic within the given set of samples. */
+    private boolean isVariant(final Genotype... gts) {
+        for (final Genotype gt : gts) {
+            if (gt.isCalled() && !gt.isHomRef()) return true;
+        }
+
+        return false;
+    }
+
+    /** Tests whether the alleles of the offspring are possible given the alleles of the parents. */
+    private boolean isMendelianViolation(final Genotype p1, final Genotype p2, final Genotype off) {
+        final Allele offAllele1 = off.getAllele(0);
+        final Allele offAllele2 = off.getAllele(1);
+
+        for (final Allele a1 : p1.getAlleles()) {
+            for (final Allele a2 : p2.getAlleles()) {
+                if (a1.equals(offAllele1) && a2.equals(offAllele2)) return false;
+                if (a2.equals(offAllele1) && a1.equals(offAllele2)) return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Tests whether the variant is within one of the pseudo-autosomal regions
+     */
+    private boolean isInPseudoAutosomalRegion(final String chr, final int pos) {
+        for (final Interval par : parIntervals) {
+            if (par.getContig().equals(chr) && pos >= par.getStart() && pos <= par.getEnd()) return true;
+        }
+
+        return false;
+    }
+
+    /** Represents the result of the work this class does. */
+    static class Result {
+        private final Collection<MendelianViolationMetrics> metrics;
+        private final Map<String, Collection<VariantContext>> violations;
+
+        Result(final Collection<MendelianViolationMetrics> metrics, final Map<String, Collection<VariantContext>> violations) {
+            this.metrics = metrics;
+            this.violations = violations;
+        }
+        
+        public Collection<MendelianViolationMetrics> metrics() {
+            return metrics;
+        }
+
+        Map<String, Collection<VariantContext>> violations() {
+            return violations;
+        }
+
+        public static MendelianViolationDetector.Result merge(final Collection<MendelianViolationDetector.Result> results) {
+            final Collection<Collection<MendelianViolationMetrics>> metricCollections = new ArrayList<>();
+            final Collection<Map<String, Collection<VariantContext>>> violationCollections = new ArrayList<>();
```

Comment:
This makes my eyes water a bit!  How about defining a simple class:
```java
class ViolationsByFamily extends HashMap<String, Collection<VariantContext>>
```
to make it clearer what's going on?